Better support ProcessPoolExecutors

mrocklin opened the pull request:
Previously we sent non-serializable state, like locks and the worker itself into the function that was sent to the executor

Now we only do this if we think that the executor is a ThreadPoolExecutor

- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `black distributed` / `flake8 distributed` / `isort distributed`

mrocklin commented:
For anyone who reviews this, I recommend going commit-by-commit.  Things should be a little simpler that way.

fjetter commented on distributed/worker.py:
```diff
@@ -2947,21 +2947,36 @@ async def execute(self, key, report=False):
-                        kwargs2,
-                        self.execution_state,
+                e = self.executors[executor]
+                if "ThreadPoolExecutor" in str(type(e)):
```
Why not an `isinstance` check?

mrocklin replied:
We've forked the `concurrent.futures.ThreadPoolExecutor` internally to support secede/rejoin.  I would want this check to work with either ours or the standard module's

crusaderky replied:
``isinstance(inst, (cls1, cls2))``?

mrocklin replied:
Loky is another example of someone implementing something like this (although they only do ProcessPoolExecutor).

To me this choice feels a bit ugly, but also ok?

fjetter reviewed (commented):
I'm quite happy about these changes since they actually reduce complexity (mostly by getting rid of the executor_submit)

The only thing missing is that the worker heartbeat no longer includes information about the executing tasks. We currently use this on scheduler side to estimate task durations.

The `active_threads` are used for profiling and we cannot simply remove them. If we introduced another collection, e.g. `executing_keys` which is managed outside of the executor we wouldn't need to serialize anything and could preserve this property executor agnostic (ignoring scheduling offset/overhead of the executor, ofc)

mrocklin commented:
> The only thing missing is that the worker heartbeat no longer includes information about the executing tasks. We currently use this on scheduler side to estimate task durations.

> The active_threads are used for profiling and we cannot simply remove them. If we introduced another collection, e.g. executing_keys which is managed outside of the executor we wouldn't need to serialize anything and could preserve this property executor agnostic (ignoring scheduling offset/overhead of the executor, ofc)

Yeah, I started down that path but then paused.  I don't recall why.  It may have been just laziness :)

mrocklin commented:
I'm not yet sure how I want to handle the SystemExit situation.  We can either return the exception or reschedule the task.  Normal operation would be that the worker would go down, and we would reschedule the task elsewhere.

mrocklin commented:
> The only thing missing is that the worker heartbeat no longer includes information about the executing tasks. We currently use this on scheduler side to estimate task durations.

To be clear.  Things still worked in the common case of ThreadPoolExecutors.

> The active_threads are used for profiling and we cannot simply remove them. If we introduced another collection, e.g. executing_keys which is managed outside of the executor we wouldn't need to serialize anything and could preserve this property executor agnostic (ignoring scheduling offset/overhead of the executor, ofc)

I've now implemented this.

fjetter commented:
> I'm not yet sure how I want to handle the SystemExit situation. 

Do you mean if the process is killed externally? If I do this with a `ProcessPoolExecutor`, I receive `concurrent.futures.process.BrokenProcessPool` regardless of the used signal

> We can either return the exception or reschedule the task. Normal operation would be that the worker would go down, and we would reschedule the task elsewhere.

If a worker dies, we're rescheduling. I think this is similar. I suggest to log the error message and reschedule.

mrocklin commented:
> If a worker dies, we're rescheduling. I think this is similar. I suggest to log the error message and reschedule.

I've moved on dev-wise.  My preference at this stage would be to leave this as-is and raise an issue if you're ok with that.

fjetter commented:
> I've moved on dev-wise. My preference at this stage would be to leave this as-is and raise an issue if you're ok with that.

sure. we can merge once green

crusaderky commented on distributed/tests/test_worker.py:
```diff
@@ -508,17 +508,15 @@ async def f(dask_worker=None):
-        w = await w
+        async with Worker(s.address, executor=e) as w:
+            assert w.executor is e
+            w = await w
```
```suggestion
```
redundant with async with

mrocklin commented on distributed/worker.py:
```diff
@@ -4020,9 +4022,8 @@ async def run(server, comm, function, args=(), kwargs=None, is_coro=None, wait=T
-    def gpu_metric(worker):
-        result = yield offload(nvml.real_time)
+    async def gpu_metric(worker):
+        result = await offload(nvml.real_time)
```
@charlesbluca @quasiben 

I think that this change should be a no-op and should not affect you.  However, it's rare to see the use of `gen.coroutine` these days and so maybe this was intentional on the part of whoever wrote this (maybe me?).  I thought I would give you a heads up just in case people start to mention that gpu metrics have started to break.

charlesbluca replied:
Thanks for the heads up!
